libs/utils/trace_analysis.py

```diff
@@ -912,196 +810,354 @@ def plotCPUFrequencyResidency(self, cpus=None):
+            # time instants where CPU leaves an idle state
+            non_idle_times = cpu_idle[cpu_idle.state == NON_IDLE_STATE]
+
+            active_indexes = cpu_idle.index.values.tolist()
+
+            exist_zero = active_indexes[0] == 0.0
+            if not exist_zero:
+                active_indexes.insert(0, 0.0)
+
+            if idle_times.index[0] < non_idle_times.index[0]:
+                # First cpu_idle entry reports an enter in idle
+                if not exist_zero:
+                    # If there wasn't a 0.0 entry in the cpu_idle trace it
+                    # means that from time 0.0 until the first cpu_idle event
+                    # there was a task running, hence first value of the square
+                    # wave must be a 1
+                    active = [(i+1)%2 for i in range(len(active_indexes))]
+                else:
+                    # If we are here first the first cpu_idle event reports an
+                    # entry in idle at time 0.0, hence first value of the
+                    # square wave must be a 0
+                    active = [i%2 for i in range(len(active_indexes))]
+            else:
+                # First cpu_idle entry reports an exit from idle
+                if not exist_zero:
+                    # If there wasn't a 0.0 entry in the cpu_idle trace it
+                    # means that from time 0.0 until the first cpu_idle event
+                    # the CPU was idle, hence first value of the square wave
+                    # must be a 0
+                    active = [i%2 for i in range(len(active_indexes))]
+                else:
+                    # If we are here first the first cpu_idle event reports an
+                    # exit from idle at time 0.0, hence first value of the
+                    # square wave must be a 1
+                    active = [(i+1)%2 for i in range(len(active_indexes))]
+
+            cpu_active[cpu] = pd.Series(active, index=active_indexes)
+
+        active = pd.DataFrame(cpu_active)
+        active.fillna(method='ffill', inplace=True)
+
+        # Cluster active is the OR between the actives on each CPU
+        # belonging to that specific cluster
+        cluster_active = reduce(
+            operator.or_,
+            [cpu_active.astype(int) for _, cpu_active in
+             active.iteritems()]
+        )
+
+        cluster_active = active[active.columns[0]].astype(int)
+        for c in active.columns[1:]:
+            cluster_active |= active[c].astype(int)
+
+        return cluster_active
+
+    def _getCPUActiveSignal(self, cpu):
+        """
+        Build a square wave representing the active (i.e. non-idle) CPU time,
+        i.e.:
+            cpu_active[t] = 1 if there is a task running in the CPU at time t
+            cpu_active[t] = 0 otherwise
+
+        :param cpu: CPU ID
+        :type cpu: int
+        """
+        return self._getClusterActiveSignal([cpu])
```

Comment:
I would actually prefer to reverse this code... I mean would it not be better to have this function implementing the code
that you have in the per-cpu loop of the _getClusterActiveSignal() call?
IMO it has more sense to implement something in that function and use it in the loop of the previous one.

This means that the Cluster code depends on the CPU code... which makes also more sense...